### PR TITLE
ui: save dag exporter

### PR DESCRIPTION
**Pull request by meowcodes**

### CODE REVIEW

File: querybook/webapp/components/DataDocDAGExporter/DataDocDAGExporterGraph.tsx [hunk 1]
```diff
@@ -26,29 +28,30 @@ export const DataDocDAGExporterGraph: React.FunctionComponent<IProps> = ({
...
-            )
-        );
-    }, [convertCellToNode, savedNodes]);
+        if (nodes.length === 0) {
+            const savedPositionsById = {};
+            savedNodes.forEach((node) => {
+                savedPositionsById[node.id] = node.position;
+            });
+            const positionedSavedNodes = queryCells.map((cell) =>
+                convertCellToNode(cell, savedPositionsById[cell.id.toString()])
```
Comment (czgu) [inline querybook/webapp/components/DataDocDAGExporter/DataDocDAGExporterGraph.tsx:50]: toString should be auto applied when inside `[]`?

File: querybook/webapp/ui/FlowGraph/FlowGraph.tsx [hunk 2]
```diff
@@ -169,35 +175,40 @@ const InteractiveFlowGraph: React.FunctionComponent<IProps> = ({
         [edges, nodes, onNodesChange]
     );
 
+    console.log('nodes', nodes);
```
Comment (czgu) [inline querybook/webapp/ui/FlowGraph/FlowGraph.tsx:178]: remove

File: querybook/webapp/components/DataDocDAGExporter/DataDocDAGExporter.tsx [hunk 3]
```diff
@@ -74,11 +83,11 @@ export const DataDocDAGExporter: React.FunctionComponent<IProps> = ({
     );
 
     React.useEffect(() => {
-        const savedNodeIds = savedDAGExport.dag?.nodes?.map((node) => node.id);
+        const savedNodeIds = savedNodes.map((node) => Number(node.id));
         setGraphQueryCells(
```
[the remark this answers is not in the document]
Comment (meowcodes) [reply querybook/webapp/components/DataDocDAGExporter/DataDocDAGExporter.tsx:87]: theres no oldCells when this runs

File: querybook/webapp/components/DataDocDAGExporter/DataDocDAGExporterGraph.tsx [hunk 4]
```diff
@@ -26,29 +28,30 @@ export const DataDocDAGExporterGraph: React.FunctionComponent<IProps> = ({
...
             },
-            position: savedPosition ?? initialNodePosition,
+            position: savedPosition || initialNodePosition,
         }),
         [onDeleteCell]
     );
 
     React.useEffect(() => {
-        setNodes(queryCells.map(convertCellToNode));
+        setNodes(queryCells.map((cell) => convertCellToNode(cell)));
```
[the remark this answers is not in the document]
Comment (meowcodes) [reply querybook/webapp/components/DataDocDAGExporter/DataDocDAGExporterGraph.tsx:40]: nodes are not cells?

File: querybook/webapp/components/DataDocDAGExporter/DataDocDAGExporterGraph.tsx [hunk 5]
```diff
@@ -1,40 +1,69 @@
...
 
+    React.useEffect(() => {
+        setNodes(
+            queryCells.map((cell: IDataQueryCell) => convertCellToNode(cell))
+        );
+    }, [convertCellToNode, queryCells]);
+
+    React.useEffect(() => {
+        if (nodes.length === 0) {
+            const savedPositionsById: Record<string, XYPosition> = {};
```
Comment (czgu) [inline querybook/webapp/components/DataDocDAGExporter/DataDocDAGExporterGraph.tsx:47]:
```
const nodes = useMemo(() => ...);
```
and `nodes.length === 0`

File: querybook/webapp/components/DataDocDAGExporter/DataDocDAGExporter.tsx [hunk 6]
```diff
@@ -24,10 +26,35 @@ export const DataDocDAGExporter: React.FunctionComponent<IProps> = ({
     docId,
     isEditable,
 }) => {
-    const { dataDocCells } = useSelector((state: IStoreState) =>
-        dataDocSelectors.dataDocSelector(state, docId)
+    const dispatch = useDispatch();
+    const { dataDocCells, savedDAGExport } = useSelector(
+        (state: IStoreState) => ({
+            dataDocCells: dataDocSelectors.dataDocSelector(state, docId)
```
[the remark this answers is not in the document]
Comment (meowcodes) [reply querybook/webapp/components/DataDocDAGExporter/DataDocDAGExporter.tsx:32]: it doesnt have the meta?

File: querybook/server/datasources/datadoc.py [hunk 7]
```diff
@@ -669,3 +665,16 @@ def send_datadoc_transfer_notification(doc_id, next_owner_id, session=None):
...
+
+@register("/datadoc/<int:id>/dag_export/", methods=["GET"])
+def get_dag_export(id):
+    assert_can_read(id)
+    verify_data_doc_permission(id)
+    return logic.get_dag_export_by_data_doc_id(data_doc_id=id)
+
+
+@register("/datadoc/<int:id>/dag_export/", methods=["POST", "PUT"])
+def update_dag_export(id, dag, meta):
```
Comment (czgu) [inline querybook/server/datasources/datadoc.py:678]: nit: create_or_update_dag_export

File: querybook/webapp/hooks/dag/useGraphQueryCells.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+import React from 'react';
+import { useDrop } from 'react-dnd';
+
+import { IStoreState } from 'redux/store/types';
+import * as dataDocSelectors from 'redux/dataDoc/selector';
+
+import { IDragItem } from 'ui/DraggableList/types';
+
+export function useGraphQueryCells(docId, savedNodes) {
+    const { dataDocCells } = useShallowSelector((state: IStoreState) =>
```
Comment (czgu) [inline querybook/webapp/hooks/dag/useGraphQueryCells.ts:13]: useSelector here, use useShallowSelect for cases when you want to return an object, like:

```
useShalldowSelector(state => ({ docs: state.dataDocs, tables: state.dataSource.tables }))

File: querybook/webapp/hooks/dag/useSavedDAG.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import { useDispatch } from 'react-redux';
+
+import { fetchDAGExport, saveDAGExport } from 'redux/dataDoc/action';
+import { IStoreState } from 'redux/store/types';
+import { useShallowSelector } from 'hooks/redux/useShallowSelector';
+
+export function useSavedDAG(docId: number) {
+    const dispatch = useDispatch();
+
+    const savedDAGExport = useShallowSelector(
```
Comment (czgu) [inline querybook/webapp/hooks/dag/useSavedDAG.ts:12]: useSelector is good here

**Pull request merged**